File: tests/AdventOfCode.Tests/PuzzleSolver.cs
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

using Microsoft.Playwright;

namespace MartinCostello.AdventOfCode;

internal sealed class PuzzleSolver(IPage page)
{
    public async Task InputArgumentsAsync(IReadOnlyList<string> arguments)
    {
        var locator = page.Locator("id=arguments");

        for (int i = 0; i < arguments.Count; i++)
        {
            if (i > 0)
            {
                await locator.FocusAsync();
                await locator.PressAsync("Enter");
            }

            await locator.PressSequentiallyAsync(arguments[i]);
        }
    }

    public async Task SelectDayAsync(string day)
        => await page.SelectOptionAsync("id=day", day);

    public async Task SelectYearAsync(string year)
        => await page.SelectOptionAsync("id=year", year);

    public async Task SelectInputAsync(string content)
    {
        byte[] buffer = Encoding.UTF8.GetBytes(content);

        var files = new FilePayload()
        {
            Buffer = buffer,
            Name = "input.txt",
            MimeType = "text/plain",
        };

        await page.SetInputFilesAsync("id=resource", files);
    }

    public async Task SolveAsync()
        => await page.ClickAsync("id=solve");

    public async Task<IList<string>> SolutionsAsync()
    {
        var container = await page.QuerySelectorAsync("id=solution");
        await container!.WaitForElementStateAsync(ElementState.Visible);

        var elements = await container.QuerySelectorAllAsync("li");

        var solutions = new List<string>(elements.Count);

        foreach (var element in elements)
        {
            solutions.Add(await element.InnerTextAsync());
        }

        return solutions;
    }

    public async Task<int> VisualizationsAsync()
    {
        var container = await page.QuerySelectorAsync("id=visualization");
        await container!.WaitForElementStateAsync(ElementState.Visible);

        var visualizations = await container.QuerySelectorAllAsync("pre");
        return visualizations.Count;
    }
}

File: src/AdventOfCode.Site/scripts/ts/puzzles/PuzzleSolver.ts
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { expect } from '@jest/globals';
import { readFileSync } from 'fs';
import { join, resolve } from 'path';
import { Puzzle } from './Puzzle';

export class PuzzleSolver {
    static async solve<T extends Puzzle>(type: { new (): T }, inputs: string[] = []): Promise<T> {
        const puzzle = new type();

        const year = PuzzleSolver.toNumberWithDigits(puzzle.year, 4);
        const day = PuzzleSolver.toNumberWithDigits(puzzle.day, 2);

        let input = join(__dirname, `../../../../AdventOfCode/Input/Y${year}/Day${day}/input.txt`);
        input = resolve(input);

        puzzle.resource = readFileSync(input, 'utf-8');

        // Act
        const result = await puzzle.solve(inputs);

        // Assert
        expect(result).not.toBeNull();
        expect(result.solutions).not.toBeNull();
        expect(result.solutions.length).toBeGreaterThan(0);

        return puzzle;
    }

    private static toNumberWithDigits(value: number, digits: number): string {
        return value.toLocaleString('en-US', {
            minimumIntegerDigits: digits,
            useGrouping: false,
        });
    }
}
